test/router-luatest/reload_test.lua

```diff
@@ -0,0 +1,265 @@
+            return box.space[space_name]:insert(tuple)
+        end)
+        rawset(_G, 'get', function(space_name, key)
+            return box.space[space_name]:get(key)
+        end)
+    end)
+end)
+
+g.after_all(function()
+    g.cluster:drop()
+end)
+
+--
+-- Reload test template:
+--     1. Invoke create_router_at:
+--         * Checkout to old version;
+--         * Create and start a router;
+--     3. Test smth on the old version;
+--     4. Invoke reload_router():
+--         * Checkout to the latest version;
+--         * Reload the module.
+--     5. Test smth on the new version
+--     6. Drop a router with vtest.drop_instance
+--
+
+local function create_router_at(hash)
+    git_util.exec('checkout', {args = hash .. ' -f', dir = g.vshard_copy_path})
+    local router = vtest.router_new(g, 'router')
+    router:exec(function(vshard_path)
+        -- Force 'require' to use new directory
+        package.path = string.format('%s/?.lua;%s/?/init.lua;',
+                                     vshard_path, vshard_path)
+        for package_name in pairs(package.loaded) do
+            if package_name:startswith('vshard') then
+                package.loaded[package_name] = nil
+            end
+        end
+    end, {g.vshard_copy_path})
+    vtest.router_cfg(router, global_cfg)
+    return router
+end
+
+local function reload_router(router)
+    git_util.exec('checkout', {args = g.latest_hash .. ' -f',
+                               dir = g.vshard_copy_path})
+    router:exec(function()
+        ilt.assert_equals(ivshard.router.module_version(), 0)
+        package.loaded['vshard.router'] = nil
+        ivshard.router = require('vshard.router')
+        ilt.assert_equals(ivshard.router.module_version(), 1)
+    end)
+end
+
+local function test_basic_template(router)
+    router:exec(function()
+        -- Basic test of callrw and callro
+        local bucket_id = 1
+        ivshard.router.callrw(bucket_id, 'insert', {'test', {1, bucket_id}})
+        ivshard.router.callrw(bucket_id, 'vshard.storage.sync', {})
+        local res, err = ivshard.router.callro(bucket_id, 'get', {'test', {1}})
+        ilt.assert_equals(err, nil)
+        ilt.assert_equals(res, {1, bucket_id})
+    end)
+end
+
+g.test_basic = function(g)
+    -- Latest meaningful commit:
+    --     "router: fix reload problem with global function refs".
+    local hash = '139223269cddefe2ba4b8e9f6e44712f099f4b35'
+    local router = create_router_at(hash)
+    test_basic_template(router)
+    reload_router(router)
+    test_basic_template(router)
+    vtest.drop_instance(g, router)
+end
+
+local function test_discovery_template(g, router)
+    -- We need to wait, until all of the services are indeed reloaded or started
+    t.helpers.retrying({timeout = vtest.wait_timeout}, function()
+        if not router:grep_log('discovery_f has been started', 1024) then
+            error("Discovery haven't started yet")
+        end
+    end)
```

Comment:
Can't you use `vshard.router.info.services` instead of the logs?

Reply:
This place hangs on my PC right now. Apparently, I have more than 1024 bytes written after the discovery was started. Would be good to redo it on service info.

Reply:
We can place it right into `reload_router`, if you don't mind. We can just compare tables of services, which is cheap.